### PR TITLE
cmake refactor

### PR DESCRIPTION
**Pull request by debris**
this pull request is enhancement of #87 

##### if fixes the following
* building `unit_testsuite` in msvc
* building in `Debug` mode in msvc
* building dynamic libraries in Xcode generated project on macos
* fixes MACOSX_RPATH policy warning on macos
* `include_directories` are not included globally any more
* unifies CMake files style
  * use 4 spaces width tabs
  * variables use upper case
  * macros/functions use lower case
* separated parts of cmake logic
  * separated CMakeCompilerSettings.cmake
  * separated CMakeDependencies.cmake
* refactored FindXXX.cmake files
  * as far as i know, we do now have to add paths `usr/include` && `usr/local/include`, they should be searched by default
  * `${CMAKE_SOURCE_DIR}/win32-deps/` moved to `CMAKE_PREFIX_PATH`, so new cmake search paths are backwards compatible
  * `FindJsoncpp.cmake` don't generate any files. Instead it's using PATH_SUFFIXES to solve jsoncpp versions incompatibility.

##### changes:
* src/jsonrpccpp/CMakeLists.txt now installs all header files, if it is unwanted, we can remove them from the list by using list REMOVE_ITEM. I think it's much cleaner solution, than previous one.

**Architecture of the library itself remained unchanged, i didn't want to create any incomatibilities with previous version**

### CODE REVIEW

**Comment by debris** [comment]
do we need these two lines in .travis.yml?
```yaml
- g++ ../src/examples/simpleclient.cpp -ljsonrpccpp-client -ljsoncpp -ljsonrpccpp-common -lcurl -o sampleclient
- g++ ../src/examples/simpleserver.cpp -ljsonrpccpp-server -ljsoncpp -ljsonrpccpp-common -lmicrohttpd -o sampleserver
```

**Comment by cinemast** [comment]
Yes. It should check if linking against the library without cmake is possible, after the library has been installed.

**Comment by debris** [comment]
are these fixes ok for you?

**Comment by debris** [comment]
Great! Yes, I'm done :)

File: src/jsonrpccpp/CMakeLists.txt [hunk 1]
```diff
@@ -1,25 +1,12 @@
...
-find_package(Threads REQUIRED)
-
 file(GLOB jsonrpc_source_common common/*.c*)
 file(GLOB jsonrpc_source_client client/*.c*)
 file(GLOB jsonrpc_source_server server/*.c*)
-list(APPEND jsonrpc_source_server server/abstractserver.h)
 
+# TODO not sure why some .h were not included
+# if we don't want to include it, remove it using list REMOVE_ITEM
 file(GLOB jsonrpc_header_common common/*.h)
```
Comment (cinemast) [inline src/jsonrpccpp/CMakeLists.txt]: Files were not included beacuse they are not part of the public API. Which makes it more easy to change internal functionality. Please revert that change, or use remove for all the others.

File: src/jsonrpccpp/common/jsonparser.h [hunk 2]
```diff
@@ -10,6 +10,6 @@
 #ifndef JSONRPC_CPP_JSONPARSER_H_
 #define JSONRPC_CPP_JSONPARSER_H_
 
-#include <@JSONCPP_INCLUDE_PREFIX@/json.h>
+#include <json/json.h>
```
Comment (cinemast) [inline src/jsonrpccpp/common/jsonparser.h]: Do not remove that. It assures that this library can build against old jsoncpp 0.6.x versions.

File: cmake/FindJSONCPP.cmake [hunk 3]
```diff
@@ -1,44 +0,0 @@
...
-    ${CMAKE_SOURCE_DIR}/win32-deps/include
-    PATH_SUFFIXES jsoncpp/json json
-)
-
-IF(JSONCPP_INCLUDE_DIRS AND JSONCPP_LIBRARIES)
-    SET(JSONCPP_FOUND TRUE)
-ENDIF(JSONCPP_INCLUDE_DIRS AND JSONCPP_LIBRARIES)
-
-IF(JSONCPP_FOUND)
-    if(EXISTS "${JSONCPP_INCLUDE_DIRS}/version.h")
```
Comment (cinemast) [inline cmake/FindJSONCPP.cmake]: These checks are required for providing old jsoncpp support (version < 0.7.x). Please revert those changes.

**Comment by debris** [comment]
**changes** :
* `jsonrpc_header_server` && `jsonrpc_client_server` are used by cmake `add_library` command. This allows all header files to be included into cmake generated project (like Xcode).
* `jsonrpc_install_header_server` && `jsonrpc_install_header_client` are used by cmake `install` command and are part of public API

**Comment by debris** [comment]
I fixed issues. Can you review pull request again?

**Comment by cinemast** [comment]
After a quick review, it looks good. I will try to build on other platforms. If everything is fine. I will merge it. Thanks again for your contribution :+1:

**Comment by cinemast** [comment]
Aside that, everything looks fine. Don't forget yourself to the AUTHORS.md ;)

**Comment by debris** [comment]
done!

**Comment by cinemast** [comment]
Great! I guess a lot of people will appreciate MSVC support. Thanks again.

**Pull request merged**